### ISSUE TITLE
Decorator not working with object references

### ISSUE THREAD

**Issue by upway** [mapstruct/mapstruct#781]
Hi,

I want to user spring decorator for reference object.
It is a bug or I have something wrong in my code but problem is that when using decorator for Dto generated code call base mapping method (InspectionResultMapper2.tireToTireDto2(tire) instead my decorated method (InspectionResultMapperDecorator2.tireToTireDto2).

Example dto and model with problem

```
public class InspectionDto2 {
    private Long id;

    private TireDto2 tire;

    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public TireDto2 getTire() {
        return tire;
    }

    public void setTire(TireDto2 tire) {
        this.tire = tire;
    }
}
```

```
public class TireDto2 {

    private Long id;

    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }
}
```
Model:
```
public class Inspection implements Serializable {

    private Tire tire;

    public Tire getTire() {
        return tire;
    }

    public void setTire(Tire tire) {
        this.tire = tire;
    }
}
```

```
public class Tire implements Serializable {
    private Long id;

    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }
}
```
Mapping condig:

```
@Mapper(componentModel = "spring")
@DecoratedWith(InspectionResultMapperDecorator2.class)
public interface InspectionResultMapper2 {

    InspectionDto2 InspectionToInspectionDto2(Inspection inspection);
    TireDto2 tireToTireDto2(Tire tire);
}
```
```
public abstract class InspectionResultMapperDecorator2 implements InspectionResultMapper2 {


    @Autowired
    @Qualifier("delegate")
    private InspectionResultMapper2 delegate;

    @Override
    public TireDto2 tireToTireDto2(Tire tire) {
        //TireDto tireDto = delegate.tireToTireDto(tire);
        TireDto2 tireDto = new TireDto2();
        tireDto.setId(222L);
        return tireDto;
    }
}
```

Generated code is:

```
@Component
@Qualifier("delegate")
public class InspectionResultMapper2Impl_ implements InspectionResultMapper2 {

    @Override
    public InspectionDto2 InspectionToInspectionDto2(Inspection inspection) {
        if ( inspection == null ) {
            return null;
        }

        InspectionDto2 inspectionDto2 = new InspectionDto2();

        inspectionDto2.setId( inspection.getId() );
/* HERE CALL UNDECORADER METHOD from InspectionResultMapper2 */
        inspectionDto2.setTire( tireToTireDto2( inspection.getTire()) );

        return inspectionDto2;
    }

    @Override
    public TireDto2 tireToTireDto2(Tire tire) {
        if ( tire == null ) {
            return null;
        }

        TireDto2 tireDto2_ = new TireDto2();

        tireDto2_.setId( tire.getId() );

        return tireDto2_;
    }
}
```
When I map Inspection with method "InspectionResultMapper2.inspectionToInspectionDto2" my decorator is not called. In generated code mapper call base method (see comment).
It is possible to use decorator for reference object ?

**Comment by agudian** [comment]
Indeed, decorators are expected to be called only from outside the decorated mapper. So if you split up your mapper into two mappers (`TireMapper` and `InspectionMapper` with `@Mapper(uses=TireMapper.class, componentModel="spring"))`), it should work.

Perhaps we should make that clearer in the documentation?

**Comment by ranjitkumarbarick** [comment]
i am also getting error  for customizing   the  Dto field       please give  the   better clarity  in documentation   .here  i mention these classes  please  evaluate the  where is the my mistake . i   have done  as per your  procedure 
@Mapper
@DecoratedWith(VendorArticleViewMapperDecorator.class)
public interface VendorArticleViewMapper {
	VendorArticleViewMapper INSTANCE=Mappers.getMapper(	VendorArticleViewMapper.class);
	 public VendorArticleViewDTO vendorArticleViewTOVendorArticleViewDTO(VendorArticleView vendorArticleView);
	 public List<VendorArticleViewDTO> vendorArticleViewTOVendorArticleViewDTOs(List<VendorArticleView> vendorArticleView);
}

this the decorate class
public abstract class VendorArticleViewMapperDecorator  implements VendorArticleViewMapper {
	
	
	
	@Autowired
	@Qualifier("delegate")
	private  VendorArticleViewMapper delegate;
	 
	
	  
	@Override
	public VendorArticleViewDTO vendorArticleViewTOVendorArticleViewDTO(VendorArticleView vendorArticleView) {
		VendorArticleViewDTO vendorArticleViewDTO=delegate.vendorArticleViewTOVendorArticleViewDTO(vendorArticleView);
		System.out.println("in delegate");
		vendorArticleViewDTO.setOutDimcm(""+vendorArticleView.getDimOuter1()+"x"+vendorArticleView.getDimOuter2()+"x"+vendorArticleView.getDimOuter3());
		return vendorArticleViewDTO;
	}

**Comment by ranjitkumarbarick** [comment]
it generated  two class  that is 
package com.smythstoys.vendormanagement.application.article;

import com.smythstoys.vendormanagement.api.article.VendorArticleViewDTO;

import java.util.ArrayList;

import java.util.List;

import javax.annotation.Generated;

@Generated(

    value = "org.mapstruct.ap.MappingProcessor",

    date = "2016-06-11T12:18:46+0530",

    comments = "version: 1.1.0.Beta1, compiler: Eclipse JDT (IDE) 1.2.0.v20150514-0146, environment: Java 1.8.0_74 (Oracle Corporation)"

)

public class VendorArticleViewMapperImpl_ implements VendorArticleViewMapper {

    @Override

    public VendorArticleViewDTO vendorArticleViewTOVendorArticleViewDTO(VendorArticleView vendorArticleView) {

        if ( vendorArticleView == null ) {

            return null;
        }

        VendorArticleViewDTO vendorArticleViewDTO = new VendorArticleViewDTO();

        vendorArticleViewDTO.setArticleBarcode( vendorArticleView.getArticleBarcode() );

        vendorArticleViewDTO.setArticleId( vendorArticleView.getArticleId() );

        vendorArticleViewDTO.setArticleName( vendorArticleView.getArticleName() );

        vendorArticleViewDTO.setAvailableDate( vendorArticleView.getAvailableDate() );

        vendorArticleViewDTO.setCartonBarcode( vendorArticleView.getCartonBarcode() );

        vendorArticleViewDTO.setCont40ft( vendorArticleView.getCont40ft() );

        vendorArticleViewDTO.setCont40ftHq( vendorArticleView.getCont40ftHq() );

        if ( vendorArticleView.getDuty() != null ) {

            vendorArticleViewDTO.setDuty( vendorArticleView.getDuty() );
        }

        vendorArticleViewDTO.setMoq( vendorArticleView.getMoq() );

        vendorArticleViewDTO.setOuterCubeFt( vendorArticleView.getOuterCubeFt() );

        if ( vendorArticleView.getOuterM3() != null ) {

            vendorArticleViewDTO.setOuterM3( vendorArticleView.getOuterM3() );
        }

        vendorArticleViewDTO.setPrice( vendorArticleView.getPrice() );

        vendorArticleViewDTO.setQtyInner( vendorArticleView.getQtyInner() );

        vendorArticleViewDTO.setQtyOuter( vendorArticleView.getQtyOuter() );

        vendorArticleViewDTO.setStatus( vendorArticleView.getStatus() );

        vendorArticleViewDTO.setVendorId( vendorArticleView.getVendorId() );

        vendorArticleViewDTO.setVendorName( vendorArticleView.getVendorName() );

        return vendorArticleViewDTO;
    }

    @Override

    public List<VendorArticleViewDTO> vendorArticleViewTOVendorArticleViewDTOs(List<VendorArticleView> vendorArticleView) {

        if ( vendorArticleView == null ) {

            return null;
        }

        List<VendorArticleViewDTO> list = new ArrayList<VendorArticleViewDTO>();

        for ( VendorArticleView vendorArticleView_ : vendorArticleView ) {

            list.add( vendorArticleViewTOVendorArticleViewDTO( vendorArticleView_ ) );
        }

        return list;
    }
}

 second  one is 
package com.smythstoys.vendormanagement.application.article;

import com.smythstoys.vendormanagement.api.article.VendorArticleViewDTO;

import java.util.List;

import javax.annotation.Generated;

@Generated(

    value = "org.mapstruct.ap.MappingProcessor",

    date = "2016-06-11T12:18:46+0530",

    comments = "version: 1.1.0.Beta1, compiler: Eclipse JDT (IDE) 1.2.0.v20150514-0146, environment: Java 1.8.0_74 (Oracle Corporation)"

)

public class VendorArticleViewMapperImpl extends VendorArticleViewMapperDecorator implements VendorArticleViewMapper {

    private final VendorArticleViewMapper delegate;

    public VendorArticleViewMapperImpl() {

        this( new VendorArticleViewMapperImpl_() );
    }

    private VendorArticleViewMapperImpl(VendorArticleViewMapperImpl_ delegate) {

        this.delegate = delegate;
    }

    @Override

    public List<VendorArticleViewDTO> vendorArticleViewTOVendorArticleViewDTOs(List<VendorArticleView> vendorArticleView)  {

        return delegate.vendorArticleViewTOVendorArticleViewDTOs( vendorArticleView );
    }
}
